ImmuDB4Net/ImmuClient.cs

```diff
@@ -1363,4 +1898,156 @@ private List<Tx> buildList(ImmudbProxy.TxList txList)
+            case ImmudbProxy.SQLValue.ValueOneofCase.Ts:
+                var dateTimeArg = DateTimeOffset.FromUnixTimeMilliseconds((long)proxyValue.Ts / 1000);
+                return new SQL.SQLValue(dateTimeArg, SqlDbType.NVarChar);
+            default:
+                throw new NotSupportedException(string.Format("The proxyvalue type {0} is not supported", proxyValue.ValueCase));
+        }
+    }
+
+    /// <summary>
+    /// Executes an SQL statement against the selected database
+    /// </summary>
+    /// <param name="sqlStatement">The SQL statement</param>
+    /// <param name="parameters">a variable number of SQLParameter values</param>
+    /// <returns>A <see cref="SQL.SQLExecResult" /> object containing transaction ids and updated rows count for each transaction</returns>
+    public async Task<SQL.SQLExecResult> SQLExec(string sqlStatement, params SQLParameter[] parameters)
+    {
+        CheckSessionHasBeenOpened();
+
+        var req = new ImmudbProxy.SQLExecRequest
+        {
+            Sql = sqlStatement,
+        };
+        if (parameters != null)
+        {
+            int paramNameCounter = 1;
+            foreach (var entry in parameters)
+            {
+                if (entry == null)
+                {
+                    req.Params.Add(new NamedParam { Name = string.Format("param{0}", paramNameCounter++), Value = CreateSQLValue(null) });
+                    continue;
+                }
+                var namedParam = new NamedParam
+                {
+                    Name = string.IsNullOrEmpty(entry.Name) ? string.Format("param{0}", paramNameCounter++) : entry.Name,
+                    Value = CreateSQLValue(entry)
+                };
+                req.Params.Add(namedParam);
+            }
+        }
+        var result = await Service.SQLExecAsync(req, Service.GetHeaders(ActiveSession));
+        var sqlResult = new SQL.SQLExecResult();
+
+        foreach (var item in result.Txs)
+        {
+            if(item.Header == null)
+            {
+                continue;
+            }
+            sqlResult.Items.Add(new SQLExecResultItem { TxID = item.Header.Id, UpdatedRowsCount = item.UpdatedRows });
+        }
+        return sqlResult;
+
+    }
+
+    /// <summary>
+    /// Executes an SQL Query against the selected database
+    /// </summary>
+    /// <param name="sqlStatement"></param>
+    /// <param name="parameters"></param>
+    /// <returns>A <see cref="SQL.SQLQueryResult" /> object containing the column list and the rows with execution result</returns>
+    public async Task<SQL.SQLQueryResult> SQLQuery(string sqlStatement, params SQLParameter[] parameters)
+    {
+        CheckSessionHasBeenOpened();
+        var req = new ImmudbProxy.SQLQueryRequest
+        {
+            Sql = sqlStatement,
+        };
+        if (parameters != null)
+        {
+            int paramNameCounter = 1;
+            foreach (var entry in parameters)
```

Comment:
This code is shared between SQLExec and SQLQuery (and will be also used by sync operations) - could be extracted to separate method

Reply:
ok, refactored, created in sql the Converters class